text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+2. Merge is performed using the leaf with highest index as right and node with second to highest index as left children; result is pushed to the end of nodes queue and leaves are discarded.
+3. Step (2) is repeated until no leaves or just one leaf remains; in latter case, the last leaf is pushed to the front of the nodes queue.
+4. Right node and then left node is popped from the front of the nodes queue and merged; the result is sent to the end of the queue.
+5. Step (4) is repeated until only one node remains; this is tree root.
+
+```
+Resulting tree for metadata consisting of 5 nodes (numbered from 0 to 4):
+
+       root
+     /     \
+    *       *
+   / \     / \
+  *   0   1   2
+ / \
+3   4
+```
+
+### Digest
+
+1. Blake3 hash is computed for each chunk of modular short metadata registry.
+3. Complete Binary Merkle Tree is constructed as described above.
+4. Root hash of this tree (left) is merged with metadata descriptor blake3 hash (right); this is metadata digest.
+
+Version number and corresponding resulting metadata digest MUST be included into Signed Extensions as specified in Chain Verification section below.
+
+### Shortening
+
+For shortening, an attempt to decode transaction completely using provided metadata is performed with the same algorithm that would be used on the cold side. All chunks are associated with their leaf indices. An example of this protocol is proposed in [metadata-shortener](https://github.com/Alzymologist/metadata-shortener) that is based on [substrate-parser](https://github.com/Alzymologist/substrate-parser) decoding protocol; any decoding protocol could be used here as long as cold signer's design finds it appropriate for given security model.
+
+### Transmission
+
+Shortened metadata chunks MAY be trasmitted into cold device together with Merkle proof in its entirety or in parts, depending on memory capabilities of the cold device and it ability to reconstruct larger fraction of tree. This document does not specify the manner of transmission. The order of metadata chunks MAY be arbitrary, the only requirement is that indices of leaf nodes in Merkle tree corresponding to chunks MUST be communicated. Community MAY handle proof format standartization independently.
+
+### Offline verification
+
+The transmitted metadata chunks are hashed together with proof lemmas to obtain root that MAY be transmitted along with the rest of payload. Verification that the root transmitted with message matches with calculated root is optional; the transmitted root SHOULD NOT be used in signature, calculated root MUST be used; however, there is no mechanism to enforce this - it should be done during cold signers code audit.
+
+### Chain verification
+
+The root of metadata computed by cold device MAY be included into Signed Extensions; this way the transaction will pass as valid iff hash of metadata as seen by cold storage device is identical to consensus hash of metadata, ensuring fair signing protocol.
```

Comment:
I mean this isn't entirely correct, because below you also explain that passing `0` means that no hash is compared.

Reply:
if `0` is passed then either root was not included into Signer Extensions ("MAY be included but it is not"), or the signature would be wrong as chain expects it to not be included ("MAY be included and is but incorrectly"); thus if what MAY be IS, we have a valid iff condition, otherwise this mechanism is just not used.

Another thing is that cold signer might just cheat and keep passing a constant, but that's out of scope here I suppose.

Maybe this clarification would be nice?

```suggestion
The root of metadata computed by cold device MAY be included into Signed Extensions; if it is included, the transaction will pass as valid iff hash of metadata as seen by cold storage device is identical to consensus hash of metadata, ensuring fair signing protocol.
```